python/ray/tune/variant_generator.py

```diff
@@ -0,0 +1,275 @@
+        return a grid search or vice versa, as long as there are no cyclic
+        dependencies between unresolved values.
+
+    Finally, to support defining specs in plain JSON / YAML, grid search
+    and lambda functions can also be defined alternatively as follows:
+
+        "activation": {"grid_search": ["relu", "tanh"]}
+        "cpu": {"eval": "spec.config.num_workers"}
+    """
+    for resolved_vars, spec in _generate_variants(unresolved_spec):
+        assert not _unresolved_values(spec)
+        yield _format_vars(resolved_vars), spec
+
+
+def grid_search(values):
+    """Convenience method for specifying grid search over a value."""
+
+    return {"grid_search": values}
+
+
+_STANDARD_IMPORTS = {
+    "random": random,
+    "np": numpy,
+}
+
+_MAX_RESOLUTION_PASSES = 20
+
+
+def _format_vars(resolved_vars):
+    out = []
+    for path, value in sorted(resolved_vars.items()):
+        if path[0] in ["alg", "env", "resources"]:
+            continue  # these settings aren't usually search parameters
+        pieces = []
+        last_string = True
+        for k in path[::-1]:
+            if isinstance(k, int):
+                pieces.append(str(k))
+            elif last_string:
+                last_string = False
+                pieces.append(k)
+        pieces.reverse()
+        out.append("_".join(pieces) + "=" + str(value))
+    return ",".join(out)
+
+
+def _generate_variants(spec):
+    spec = copy.deepcopy(spec)
+    unresolved = _unresolved_values(spec)
+    if not unresolved:
+        yield {}, spec
+        return
+
+    grid_vars = []
+    lambda_vars = []
+    for path, value in unresolved.items():
+        if isinstance(value, types.FunctionType):
+            lambda_vars.append((path, value))
+        else:
+            grid_vars.append((path, value))
+    grid_vars.sort()
+
+    grid_search = _grid_search_generator(spec, grid_vars)
+    for resolved_spec in grid_search:
+        resolved_vars = _resolve_lambda_vars(resolved_spec, lambda_vars)
+        for resolved, spec in _generate_variants(resolved_spec):
+            for path, value in grid_vars:
+                resolved_vars[path] = _get_value(spec, path)
+            for k, v in resolved.items():
+                if (k in resolved_vars and v != resolved_vars[k] and
+                        _is_resolved(resolved_vars[k])):
+                    raise ValueError(
+                        "The variable `{}` could not be unambiguously "
+                        "resolved to a single value. Consider simplifying "
+                        "your variable dependencies.".format(k))
+                resolved_vars[k] = v
+            yield resolved_vars, spec
+
+
+def _assign_value(spec, path, value):
+    for k in path[:-1]:
+        spec = spec[k]
+    spec[path[-1]] = value
+
+
+def _get_value(spec, path):
+    for k in path:
+        spec = spec[k]
+    return spec
+
+
+def _resolve_lambda_vars(spec, lambda_vars):
```

Comment:
Would be nice to document this function, specifically the usage of MAX RESOLUTION PASSES